src/gui/application.cpp

```diff
@@ -551,51 +544,105 @@ namespace Toolbox {
         }
     }
 
-    void FileDialog::OpenDialog(std::filesystem::path starting_path,
-                                GLFWwindow* parent_window,
-                                bool is_directory,
-                                std::optional<std::vector<std::pair<
-                                  const char*, const char*>>> maybe_filters) {
+#ifdef TOOLBOX_PLATFORM_WINDOWS
+    void FileDialog::OpenDialog(
+        std::filesystem::path starting_path, GLFWwindow *parent_window, bool is_directory,
+        std::optional<std::vector<std::pair<std::string, std::string>>> maybe_filters) {
+        if (m_thread_initialized) {
+            m_thread.join();
+        } else {
+            m_thread_initialized = true;
+        }
+        m_thread_running = true;
+        m_closed         = false;
+        auto fn          = [this, starting_path, parent_window, is_directory, maybe_filters]() {
+            m_starting_path = starting_path.wstring();
+            m_filters.clear();
+            if (is_directory) {
+                nfdpickfoldernargs_t args;
+                args.defaultPath = m_starting_path.c_str();
+                NFD_GetNativeWindowFromGLFWWindow(parent_window, &args.parentWindow);
+                m_result = NFD_PickFolderN_With(&m_selected_path, &args);
+                TOOLBOX_INFO_V("Selected path: {}", std::filesystem::path(m_selected_path).string());
+                TOOLBOX_INFO_V("Result: {}", int(m_result));
+                // m_result = NFD_PickFolderN(&m_selected_path, starting_path.string().c_str());
+            } else {
+                int num_filters               = 0;
+                nfdnfilteritem_t *nfd_filters = nullptr;
+                if (maybe_filters) {
+                    auto filters = maybe_filters.value();
+                    num_filters  = filters.size();
+                    nfd_filters  = new nfdnfilteritem_t[num_filters];
+                    for (int i = 0; i < num_filters; ++i) {
+                        std::wstring filter_name =
+                            std::wstring(filters[i].first.begin(), filters[i].first.end());
+                        std::wstring filter_ext =
+                            std::wstring(filters[i].second.begin(), filters[i].second.end());
+                        m_filters.push_back({filter_name, filter_ext});
+                        nfd_filters[i] = {m_filters[i].first.c_str(), m_filters[i].second.c_str()};
```

Comment:
Huh, how is this well-typed? c_str returns a char*, but nfdfilteritem_t is a wchar_t*, right? Is there an implicit conversion going on here?

Reply:
Oh wait, m_filters holds wstrings, so maybe that means that c_str returns a wchar_t* for them, nevermind.